lib/remote_node.go

```diff
@@ -0,0 +1,648 @@
+	}
+
+	if rn.GetPeer().IsOutbound() {
+		versionTimeExpected := time.Now().Add(rn.params.VersionNegotiationTimeout)
+		rn.versionTimeExpected = &versionTimeExpected
+		if err := rn.sendVersionMessage(nonce); err != nil {
+			return fmt.Errorf("InitiateHandshake: Problem sending version message to peer (id= %d): %v", rn.id, err)
+		}
+		rn.setHandshakeStage(HandshakeStage_VersionSent)
+	}
+	return nil
+}
+
+// sendVersionMessage generates and sends a version message to a RemoteNode peer. The message will contain the nonce
+// that is passed in as an argument.
+func (rn *RemoteNode) sendVersionMessage(nonce uint64) error {
+	verMsg := rn.newVersionMessage(nonce)
+
+	// Record the nonce of this version message before we send it so we can
+	// detect self connections and so we can validate that the peer actually
+	// controls the IP she's supposedly communicating to us from.
+	rn.handshakeMetadata.versionNonceSent = nonce
+
+	if err := rn.SendMessage(verMsg); err != nil {
+		return fmt.Errorf("sendVersionMessage: Problem sending version message to peer (id= %d): %v", rn.id, err)
+	}
+	rn.setHandshakeStage(HandshakeStage_VersionSent)
+	return nil
+}
+
+// newVersionMessage returns a new version message that can be sent to a RemoteNode peer. The message will contain the
+// nonce that is passed in as an argument.
+func (rn *RemoteNode) newVersionMessage(nonce uint64) *MsgDeSoVersion {
+	ver := NewMessage(MsgTypeVersion).(*MsgDeSoVersion)
+
+	ver.Version = rn.params.ProtocolVersion.ToUint64()
+	// Set the services bitfield to indicate what services this node supports.
+	ver.Services = rn.nodeServices
+
+	// We use an int64 instead of a uint64 for convenience.
+	ver.TstampSecs = time.Now().Unix()
+
+	ver.Nonce = nonce
+	ver.UserAgent = rn.params.UserAgent
+
+	// When a node asks you for what height you have, you should reply with the height of the latest actual block you
+	// have. This makes it so that peers who have up-to-date headers but missing blocks won't be considered for initial
+	// block download.
+	ver.LatestBlockHeight = rn.latestBlockHeight
+
+	// Set the minimum fee rate the peer will accept.
+	ver.MinFeeRateNanosPerKB = rn.minTxFeeRateNanosPerKB
+
+	return ver
+}
+
+// HandleVersionMessage is called upon receiving a version message from the RemoteNode's peer. The peer may be the one
+// initiating the handshake, in which case, we should respond with our own version message. To do this, we pass the
+// responseNonce to this function, which we will use in our response version message.
+func (rn *RemoteNode) HandleVersionMessage(verMsg *MsgDeSoVersion, responseNonce uint64) error {
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	if rn.connectionStatus != RemoteNodeStatus_Connected {
+		return fmt.Errorf("HandleVersionMessage: RemoteNode is not connected")
+	}
+	if rn.getHandshakeStage() != HandshakeStage_NotStarted && rn.getHandshakeStage() != HandshakeStage_VersionSent {
```

Comment:
This should also be an OR, no? 
```suggestion
	if rn.getHandshakeStage() != HandshakeStage_NotStarted || rn.getHandshakeStage() != HandshakeStage_VersionSent {
```

Reply:
Hmm, I think it should be AND. We should continue this function if NotStarted || VersionSent, and error if the negation is true. So the De Morgan is !NotStarted && !VersionSent.